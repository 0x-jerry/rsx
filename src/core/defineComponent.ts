import { isFn } from '@0x-jerry/utils'
import type { Merge } from 'type-fest'
import type { ToMaybeRef } from '.'
import type { DefineProps } from './props'

export interface PropOption<T = any> {
  type?: T
  default?: T | (() => T)
}

export type PropsType<T = any> = Record<string, T | PropOption<T>>

export interface ComponentOption<Props extends PropsType = {}> {
  props?: Props
  toRefs?: boolean
}

interface CommonProps {
  class?: any
  [key: string]: any
}

export type FunctionalComponent<P extends PropsType = {}> = (
  props: Merge<P, CommonProps>,
  children?: any[],
) => JSX.Element

/**
 * without option
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * with option
 * @param opt
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P>,
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * implement
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P> | FunctionalComponent,
  impl?: FunctionalComponent,
): FunctionalComponent {
  if (isFn(opt)) return opt

  // todo check props type
  return impl!
}

export const dc = defineComponent
