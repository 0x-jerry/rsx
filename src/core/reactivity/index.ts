export { computed, reactive, ref } from '@vue/reactivity'
export * from './effect'
export { nextTick } from './scheduler'
export * from './types'
export * from './utils'
export * from './watch'
