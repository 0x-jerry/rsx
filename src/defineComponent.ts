import type { EmptyObject } from '@0x-jerry/utils'
import type { DefineProps } from './props'
import type { IsStartWithCapitalizedLetter, MaybeRef } from './types'

export type PropsType<T = any> = Record<string, T>

type ToPropRefs<T extends {}> = {
  /**
   * Start with capitalized letter means this prop is not used as reactive prop
   */
  [key in keyof T]: key extends IsStartWithCapitalizedLetter<key>
    ? T[key]
    : MaybeRef<T[key]>
}

/**
 * @private
 */
export type FunctionalComponent<P extends PropsType = any> = (
  props: P,
  children?: any[],
) => unknown

type _NoInferProps<P extends EmptyObject> = DefineProps<P> & Record<string, any>

export type ExposedFunctionalComponent<P extends PropsType = any> =
  FunctionalComponent<ToPropRefs<P>>

/**
 * Auto calculate props from generic type P
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: NoInfer<FunctionalComponent<_NoInferProps<P>>>,
): ExposedFunctionalComponent<P>

/**
 * Infer props from impl parameter
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<P>,
): ExposedFunctionalComponent<P>

/**
 * implement
 */
export function defineComponent(impl?: any): any {
  // todo check props type
  return impl
}

export const dc = defineComponent
