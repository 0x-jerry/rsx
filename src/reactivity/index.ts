export { computed, reactive, ref } from '@vue/reactivity'
export * from './bindingRef'
export * from './effect'
export { nextTick } from './scheduler'
export * from './types'
export * from './watch'
