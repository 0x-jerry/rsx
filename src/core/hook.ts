import type { Fn } from '@0x-jerry/utils'
import type { WatchCallback, WatchEffect, WatchSource } from '@vue/reactivity'
import {
  type DNodeContext,
  DNodeContextEventName,
  getCurrentContext,
} from './context'
import { type WatchHandle, type WatchOptions, watch } from './reactivity'

export function unmount(ctx: DNodeContext) {
  ctx.emit(DNodeContextEventName.beforeUnmount)

  ctx.el?.remove()

  ctx.children?.forEach((child) => unmount(child))

  ctx.emit(DNodeContextEventName.unmounted)
}

export function mount(ctx: DNodeContext) {
  ctx.emit(DNodeContextEventName.beforeMount)

  ctx.children?.forEach((item) => mount(item))

  ctx.emit(DNodeContextEventName.mounted)
}

export function useContext() {
  const ctx = getCurrentContext()

  if (!ctx) {
    throw new Error('This should only used inside functional component.')
  }

  return ctx
}

export function onBeforeMount(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.beforeMount, fn)
}

export function onMounted(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.mounted, fn)
}

export function onBeforeUnmount(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.beforeUnmount, fn)
}

export function onUnmounted(fn: Fn, ctx?: DNodeContext) {
  ctx ||= useContext()

  ctx.on(DNodeContextEventName.unmounted, fn)
}

export function useWatch(
  getter: WatchSource | WatchSource[] | WatchEffect,
  fn: WatchCallback,
  opt?: WatchOptions,
): WatchHandle {
  const stop = watch(getter, fn, opt)

  onUnmounted(stop)

  return stop
}
