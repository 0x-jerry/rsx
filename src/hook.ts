import type { EmptyObject, Fn } from '@0x-jerry/utils'
import {
  readonly,
  type WatchCallback,
  type WatchEffect,
  type WatchSource,
} from '@vue/reactivity'
import {
  type DNodeContext,
  DNodeContextEventName,
  type DNodeEventMap,
  getCurrentContext,
} from './context'
import { isRef, type WatchHandle, type WatchOptions, watch } from './reactivity'

export function useContext() {
  const ctx = getCurrentContext()

  if (!ctx) {
    throw new Error('This should only used inside functional component.')
  }

  return ctx
}

function createContextHook(name: keyof DNodeEventMap) {
  return (fn: Fn, ctx?: DNodeContext) => {
    ctx ||= useContext()

    ctx.on(name, fn)
  }
}

export const onBeforeMount = createContextHook(
  DNodeContextEventName.beforeMount,
)
export const onMounted = createContextHook(DNodeContextEventName.mounted)

export const onBeforeUnmount = createContextHook(
  DNodeContextEventName.beforeUnmount,
)
export const onUnmounted = createContextHook(DNodeContextEventName.unmounted)

export function useWatch(
  getter: WatchSource | WatchSource[] | WatchEffect,
  fn: WatchCallback,
  opt?: WatchOptions,
): WatchHandle {
  const stopHandle = watch(getter, fn, opt)

  onUnmounted(stopHandle)

  return stopHandle
}

export type InjectionKey<T> = (string | symbol) & { _: T }

export function provide<T = unknown>(
  key: string | symbol | InjectionKey<T>,
  value: T,
) {
  const ctx = useContext()
  ctx.ex ||= {}
  ctx.ex[key] = value
}

export function inject<T>(
  key: string | symbol | InjectionKey<T>,
): T | undefined {
  const ctx = useContext()

  let node = ctx.parent

  while (node) {
    if (node.ex?.[key]) {
      return node.ex[key] as T
    }

    node = node.parent
  }
}

export function useRawProps() {
  const ctx = useContext()

  return ctx._node?.props || {}
}

export function useExpose<T extends EmptyObject>(exposed: T) {
  const ref = useRawProps().ref

  if (isRef(ref)) {
    ref.value = readonly(exposed)
  }
}

export function useRawChildren() {
  return useContext()._node?.children || []
}
