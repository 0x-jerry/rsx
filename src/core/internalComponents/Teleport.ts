import { unref } from '@vue/reactivity'
import { onMounted, useWatch } from '../hook'
import { createTextElement } from '../node'
import { moveChildren } from '../nodeOp'

export function Teleport(
  props: {
    to: string
  },
  children?: any[],
) {
  const el = createTextElement('')

  useWatch(() => unref(props.to), update)

  onMounted(update)

  return el as unknown as JSX.Element

  function update() {
    const selector = unref(props.to)
    const root = document.querySelector(selector)

    if (!root) {
      return
    }

    moveChildren(root, children)
  }
}
