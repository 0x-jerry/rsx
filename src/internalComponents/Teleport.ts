import { runWithContext } from '@/context'
import { moveTo } from '@/nodeOp'
import { normalizeNodes } from '@/nodes/shared'
import { defineComponentName } from '@/test'
import { defineComponent } from '../defineComponent'
import { onBeforeMount, useContext, useWatch } from '../hook'

export interface TeleportProps {
  to?: string
}

export const Teleport = defineComponent<TeleportProps>((props, children) => {
  const ctx = useContext()

  if (ctx._node) {
    ctx._node.children = normalizeNodes(children || [])
  }

  let initialized = false

  const _update = () => runWithContext(update, ctx)

  useWatch(() => props.to, _update)
  onBeforeMount(_update)

  const nonExistsContainer = document.createDocumentFragment()

  function update() {
    const container = props.to
      ? document.querySelector(props.to) || nonExistsContainer
      : nonExistsContainer

    if (!initialized) {
      for (const child of ctx._node?.children || []) {
        child.initialize()
        if (child.el) {
          moveTo(container, child.el)
        }
      }
      initialized = true
    } else {
      for (const child of ctx._node?.children || []) {
        if (child.el) {
          moveTo(container, child.el)
        }
      }
    }
  }

  return null
})

defineComponentName(Teleport, 'Teleport')
