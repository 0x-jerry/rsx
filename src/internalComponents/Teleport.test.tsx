import { onMounted } from '@/hook'
import { nextTick, ref } from '@/reactivity'
import { dc } from '../defineComponent'
import { mountTestApp } from '../test'
import { Teleport } from './Teleport'

describe('Teleport', () => {
  it('non-reactivity data', async () => {
    const A = () => (
      <div>
        <span>0</span>
        <div id="tp"></div>
      </div>
    )
    const App = dc(() => {
      return (
        <>
          <span>1</span>
          <A></A>
          <Teleport to="#tp">
            <span>2</span>
          </Teleport>
        </>
      )
    })

    const root = mountTestApp(App)

    expect(root.outerHTML).toMatchSnapshot()
  })

  it('reactivity data', async () => {
    const A = () => (
      <div>
        <span>0</span>
        <div id="tp"></div>
        <div id="tp2"></div>
      </div>
    )
    const App = dc(() => {
      const toValue = ref('#tp')

      onMounted(() => {
        toValue.value = '#tp2'
      })

      return (
        <>
          <span>1</span>
          <A></A>
          <Teleport to={toValue}>
            <span>2</span>
          </Teleport>
        </>
      )
    })

    const root = mountTestApp(App)
    await nextTick()

    expect(root.outerHTML).toMatchSnapshot()
  })

  it('container node is not exists', async () => {
    const A = () => (
      <div>
        <span>0</span>
        <div id="tp"></div>
        <div id="tp2"></div>
      </div>
    )
    const App = dc(() => {
      const toValue = ref('#tp')

      onMounted(() => {
        toValue.value = '#tp3'
      })

      return (
        <>
          <span>1</span>
          <A></A>
          <Teleport to={toValue}>
            <span>2</span>
          </Teleport>
        </>
      )
    })

    const root = mountTestApp(App)
    await nextTick()

    expect(root.outerHTML).toMatchSnapshot()
  })
})
