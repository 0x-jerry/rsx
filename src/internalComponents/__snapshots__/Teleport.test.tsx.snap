// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Teleport > container node is not exists > html 1`] = `
<div
  id="app"
>
  <span>
    1
  </span>
  <div>
    <span>
      0
    </span>
    <div
      id="tp"
    />
    <div
      id="tp2"
    />
  </div>
  <!--fragment-->
</div>
`;

exports[`Teleport > non-reactivity data > html 1`] = `
<div
  id="app"
>
  <span>
    1
  </span>
  <div>
    <span>
      0
    </span>
    <div
      id="tp"
    >
      <span>
        2
      </span>
    </div>
  </div>
  <!--fragment-->
</div>
`;

exports[`Teleport > reactivity data > html 1`] = `
<div
  id="app"
>
  <span>
    1
  </span>
  <div>
    <span>
      0
    </span>
    <div
      id="tp"
    />
    <div
      id="tp2"
    >
      <span>
        2
      </span>
    </div>
  </div>
  <!--fragment-->
</div>
`;
